app/src/main/java/com/github/se/travelpouch/model/activity/ActivityRepositoryFirebase.kt

```diff
@@ -0,0 +1,160 @@
+        Log.d("EventRepository", "getEvents")
+        db.collection(collectionPath)
+            .get()
+            .addOnSuccessListener { result ->
+                val events = result?.mapNotNull { documentToEvent(it) } ?: emptyList()
+                onSuccess(events)
+            }
+            .addOnFailureListener { e ->
+                Log.e("EventRepository", "Error getting documents", e)
+                onFailure(e)
+            }
+    }
+
+    /**
+     * This function adds an event to the collection of events in Firebase.
+     *
+     * @param event (Event) : the event we want to add on Firebase
+     * @param onSuccess (() -> Unit) : the function called when the event is correctly added to the
+     *   database
+     * @param onFailure ((Exception) -> Unit) : the function called when an error occurs during the
+     *   adding an event to the database
+     */
+    override fun addActivity(
+        activity: Activity,
+        onSuccess: () -> Unit,
+        onFailure: (Exception) -> Unit
+    ) {
+        performFirestoreOperation(
+            db.collection(collectionPath).document(activity.uid).set(activity), onSuccess, onFailure)
+    }
+
+    /**
+     * This function is a helper function that safely performs a Firebase operation. A task has
+     * listeners added to it. If the task is successful, we apply onSuccess. Otherwise we perform
+     * onFailure.
+     *
+     * @param task (Task<Void>) : a task to perform
+     * @param onSuccess (() -> Unit) : the function called when the event is correctly added to the
+     *   database
+     * @param onFailure ((Exception) -> Unit) : the function called when an error occurs during the
+     *   adding an event to the database
+     */
+    private fun performFirestoreOperation(
+        task: Task<Void>,
+        onSuccess: () -> Unit,
+        onFailure: (Exception) -> Unit
+    ) {
+        task
+            .addOnSuccessListener { onSuccess() }
+            .addOnFailureListener { e ->
+                Log.e("EventRepositoryFirestore", "Error performing Firestore operation", e)
+                onFailure(e)
+            }
+    }
+
+    /**
+     * This function converts a document got from Firebase to an event. It returns null if an error
+     * occurs.
+     *
+     * @param document (DocumentSnapshot) : The document from Firebase
+     * @return (Event?) : If the conversion goes without a problem an event is return. Otherwise, null
+     *   is returned.
+     */
+    private fun documentToEvent(document: DocumentSnapshot): Activity? {
+        return try {
+            val uid = document.id
+            val title = document.getString("title")
+            val description = document.getString("description")
+            val date = document.getTimestamp("date")
+            val documents = document.get("documentsNeeded") as? Map<String, Int>
+            val locationData = document.get("location") as? Map<*, *>
```

Comment:
It would be better to do it like that for documents and locationData to provides safer type checking and conversion:
 ```kotlin
val documentsNeededData = document["documentsNeeded"] as? Map<*, *>
val documentsNeeded = documentsNeededData?.map { (key, value) -> key as String to value as Int }?.toMap()

Reply:
Yes, I agree. I did know this notation